openhtf/io/frontend/src/app/test.ts

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016 Google Inc.All Rights Reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+/** Represents an active or past test run. **/
+export class Test {
+  dutID: string;
+  startTime: number;
+  endTime: number;
+  duration: number;
+  outcome: string;
+  phases: Phase[];
+  logEntries: LogEntry[];
+
+  /**
+   * Create a Test.
+   * @param record - A serialized OpenHTF TestRecord.
+   */
+  constructor() {
+    this.duration = 0;
+    this.logEntries = [];
+    this.phases = [];
+  }
+
+  /**
+   * Update state based on a test record.
+   * @param test - A serialized OpenHTF TestState.
+   * @param phaseList - An option list of phases to override the record.
+   */
+  update(record: any) {
+    // Test state is tricky to track and update particularly because we want
+    // log entries and other data displayed in the gui to be selectable. If
+    // new components are being created on every update, we will lost active
+    // selections even when the gui appears to remain static. Instead, we need
+    // to keep the gui _actually_ static when no changes are happening, which
+    // means the inputs to view componts need to stay pointing at the same
+    // objects in memory. If someone has a better way to do this, please let
```

Comment:
Here's one option that could even lower the bandwidth of this UI:

The server currently sends the whole test state each time. If it stored the previous N states with a monotonically increasing number, the frontend would ask for an update and pass it it's current state number. The server would return a json patch object (http://jsonpatch.com/) that would be able to convert state #X to #Y, assuming #X was in the server's history. If not, it would return the state as-is and the UI would replace everything, but that would be a rare case of slow-bandwidth and the UI has probably changes significantly in the meantime.

This would 'solve' the problem by taking advantage of json patches not replacing objects, only adding/replacing/removing/etc, which is what all of this code does. It also reduces the number of times the test state organization is duplicated (by the framework, the typescript, and the HTML at the moment) since this doesn't care about the organization anymore.